File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/common/Robot.java
package org.firstinspires.ftc.teamcode.common;

import android.content.Context;
import android.hardware.Sensor;
import android.hardware.SensorEvent;
import android.hardware.SensorEventListener;
import android.hardware.SensorManager;

import com.qualcomm.hardware.bosch.BNO055IMU;
import com.qualcomm.hardware.bosch.JustLoggingAccelerationIntegrator;
import com.qualcomm.robotcore.hardware.ColorSensor;
import com.qualcomm.robotcore.hardware.DcMotor;
import com.qualcomm.robotcore.hardware.DcMotorEx;
import com.qualcomm.robotcore.hardware.HardwareMap;

import org.firstinspires.ftc.robotcontroller.external.samples.SensorREVColorDistance;
import org.firstinspires.ftc.robotcore.external.Telemetry;
import org.firstinspires.ftc.robotcore.external.navigation.Orientation;

public class Robot {

    public static final double WHEEL_DISTANCE_INCHES = 18.1;
    public static final double MAX_NEVE_VELOCITY = 3120.0;

    public static final double     NEVE_COUNTS_PER_MOTOR_NEVE = 1120;      // eg: Rev Side motor
    public static final double            DRIVE_GEAR_REDUCTION    = 45.0 / 35.0;             // This is < 1.0 if geared UP
    public static final double            WHEEL_DIAMETER_INCHES   = 5.250;           // For figuring circumference
    public static final double     COUNTS_PER_INCH = (NEVE_COUNTS_PER_MOTOR_NEVE * DRIVE_GEAR_REDUCTION) /
            (WHEEL_DIAMETER_INCHES * Math.PI);

    public final DcMotorEx fl;
    public final DcMotorEx fr;
    public final DcMotorEx bl;
    public final DcMotorEx br;

    public Orientation angles = new Orientation();

    public final BNO055IMU imu;
    public final BNO055IMU.Parameters params;

    private Telemetry telemetry;

    private boolean isReverse = false;

    public Robot(HardwareMap hardwareMap, Telemetry telemetry) {
        this.telemetry = telemetry;
        fl = hardwareMap.get(DcMotorEx.class, "fl");
        fr = hardwareMap.get(DcMotorEx.class, "fr");
        bl = hardwareMap.get(DcMotorEx.class, "bl");
        br = hardwareMap.get(DcMotorEx.class, "br");

        fl.setZeroPowerBehavior(DcMotor.ZeroPowerBehavior.BRAKE);
        fr.setZeroPowerBehavior(DcMotor.ZeroPowerBehavior.BRAKE);
        bl.setZeroPowerBehavior(DcMotor.ZeroPowerBehavior.BRAKE);
        br.setZeroPowerBehavior(DcMotor.ZeroPowerBehavior.BRAKE);

        fr.setDirection(DcMotorEx.Direction.REVERSE);
        br.setDirection(DcMotorEx.Direction.REVERSE);

        params = new BNO055IMU.Parameters();

        params.angleUnit           = BNO055IMU.AngleUnit.DEGREES;
        params.accelUnit           = BNO055IMU.AccelUnit.METERS_PERSEC_PERSEC;
        params.calibrationDataFile = "BNO055IMUCalibration.json"; // see the calibration sample opmode
        params.loggingEnabled      = true;
        params.loggingTag          = "IMU";
        params.accelerationIntegrationAlgorithm = new JustLoggingAccelerationIntegrator();

        imu = hardwareMap.get(BNO055IMU.class, "imu");
    }

    public void setMode(DcMotor.RunMode runMode) {
        fl.setMode(runMode);
        fr.setMode(runMode);
        bl.setMode(runMode);
        br.setMode(runMode);
    }

    public void setReverse(boolean value) {
        if(value) {
            fl.setDirection(DcMotorEx.Direction.REVERSE);
            bl.setDirection(DcMotorEx.Direction.REVERSE);
            fr.setDirection(DcMotorEx.Direction.FORWARD);
            br.setDirection(DcMotorEx.Direction.FORWARD);
        } else {
            fl.setDirection(DcMotorEx.Direction.FORWARD);
            bl.setDirection(DcMotorEx.Direction.FORWARD);
            fr.setDirection(DcMotorEx.Direction.REVERSE);
            br.setDirection(DcMotorEx.Direction.REVERSE);
        }
        isReverse = value;
    }

    public boolean isReverse() {
        return isReverse;
    }

    public boolean isBusy() {
        return fl.isBusy() || fr.isBusy() || bl.isBusy() || br.isBusy();
    }

}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/common/Robot.kt
package org.firstinspires.ftc.teamcode.common

import com.qualcomm.robotcore.hardware.DcMotor
import com.qualcomm.robotcore.hardware.DcMotorEx
import com.qualcomm.robotcore.hardware.DcMotorSimple
import com.qualcomm.robotcore.hardware.HardwareMap

class Robot(hardwareMap: HardwareMap) {
    val WHEEL_DISTANCE_INCHES = 18.1

    val NEVE_COUNTS_PER_MOTOR_NEVE = 1120.0      // eg: Rev Side motor
    val DRIVE_GEAR_REDUCTION = 45.0 / 35.0             // This is < 1.0 if geared UP
    val WHEEL_DIAMETER_INCHES = 5.250           // For figuring circumference
    val COUNTS_PER_INCH = NEVE_COUNTS_PER_MOTOR_NEVE * DRIVE_GEAR_REDUCTION / (WHEEL_DIAMETER_INCHES * Math.PI)

    val fl: DcMotorEx
    val fr: DcMotorEx
    val bl: DcMotorEx
    val br: DcMotorEx

    private var isReverse = false

    init {
        fl = hardwareMap.get("fl") as DcMotorEx
        fr = hardwareMap.get("fr") as DcMotorEx
        bl = hardwareMap.get("bl") as DcMotorEx
        br = hardwareMap.get("br") as DcMotorEx

        fl.zeroPowerBehavior = DcMotor.ZeroPowerBehavior.BRAKE
        fr.zeroPowerBehavior = DcMotor.ZeroPowerBehavior.BRAKE
        bl.zeroPowerBehavior = DcMotor.ZeroPowerBehavior.BRAKE
        br.zeroPowerBehavior = DcMotor.ZeroPowerBehavior.BRAKE

        fr.direction = DcMotorSimple.Direction.REVERSE
        br.direction = DcMotorSimple.Direction.REVERSE
    }

    fun setMode(runMode: DcMotor.RunMode) {
        fl.mode = runMode
        fr.mode = runMode
        bl.mode = runMode
        br.mode = runMode
    }

    fun setReverse(value: Boolean) {
        if (value) {
            fl.direction = DcMotorSimple.Direction.REVERSE
            bl.direction = DcMotorSimple.Direction.REVERSE
            fr.direction = DcMotorSimple.Direction.FORWARD
            br.direction = DcMotorSimple.Direction.FORWARD
        } else {
            fl.direction = DcMotorSimple.Direction.FORWARD
            bl.direction = DcMotorSimple.Direction.FORWARD
            fr.direction = DcMotorSimple.Direction.REVERSE
            br.direction = DcMotorSimple.Direction.REVERSE
        }
        isReverse = value
    }

    fun isReverse() : Boolean {
        return isReverse
    }

    fun isBusy() : Boolean {
        return fl.isBusy || fr.isBusy || bl.isBusy || br.isBusy
    }

}